### PR TITLE
fix: change the current product criteria for sync

### PR DESCRIPTION
**Pull request by leogermani**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Updates the criteria to get what is considered the "Current product" for sync purposes

### How to test the changes in this Pull Request:

1. Just for testing, make the `get_current_product_order_for_sync` method public
2. Create a user with different sets of purchases, and make sure the method always return the expected result, according to tis description
3. Make sure to test subscriptions, expired subscriptions, one time donations, recurring donations and a user with only a regular purchase (not subscription and not donation)

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208262205728470

### CODE REVIEW

File: includes/reader-activation/sync/class-woocommerce.php [hunk 1]
```diff
@@ -40,6 +40,118 @@ public static function should_sync_order( $order ) {
...
+	 * @param int $user_id User ID.
+	 *
+	 * @return ?WCS_Subscription A Subscription object or null.
+	 */
+	private static function get_most_recent_cancelled_or_expired_subscription( $user_id ) {
+		$subcriptions = array_reduce(
+			array_keys( \wcs_get_users_subscriptions( $user_id ) ),
+			function( $acc, $subscription_id ) {
+				$subscription = \wcs_get_subscription( $subscription_id );
+				if ( $subscription->has_status( [ 'on-hold', 'cancelled', 'expired' ] ) ) {
```
Comment (dkoo) [inline includes/reader-activation/sync/class-woocommerce.php]: Nit: to keep things consistent, should we use [this constant](https://github.com/Automattic/newspack-plugin/blob/fix/current-product-criteria/includes/reader-revenue/woocommerce/class-woocommerce-connection.php#L19)? e.g.

```suggestion
				if ( ! $subscription->has_status( WooCommerce_Connection::ACTIVE_SUBSCRIPTION_STATUSES ) ) {
```

File: tests/unit-tests/reader-activation-sync.php [hunk 2]
```diff
@@ -74,8 +74,6 @@ public function test_can_esp_sync() {
 		Reader_Activation::update_setting( 'mailchimp_audience_id', '123' );
 		$errors = ESP_Sync::can_esp_sync( true );
 		$this->assertNotContains( 'ras_esp_master_list_id_not_found', $errors->get_error_codes(), 'Master list ID is set' );
-
-		$this->assertTrue( ESP_Sync::can_esp_sync(), 'Reader data should be syncable after conditions are met' );
```
Comment (dkoo) [inline tests/unit-tests/reader-activation-sync.php]: Did you mean to remove this test?

**Review by dkoo — CHANGES_REQUESTED** [review]
Works as described! A minor nitpick, and a non-blocking question below. Also, do you think it's worth adding a unit test where we mock some users who own various types of products (and some who own none), to test the logic? Or is that overkill?

File: tests/unit-tests/reader-activation-sync.php (hunk 2 above)
Comment (leogermani) [reply tests/unit-tests/reader-activation-sync.php]: Yes, this test won't pass and was already broken in trunk. Because it checks for `class_exists( 'Newspack_Newsletters_Contacts' ) `.

But we can just declare a mock class instead as well...

File: includes/reader-activation/sync/class-woocommerce.php (hunk 1 above)
Comment (leogermani) [reply includes/reader-activation/sync/class-woocommerce.php]: That was actually intentional, but open for discussion.

My idea is that maybe we don't want to consider all non-active statuses there. Because what we are looking for are former donors/subscribers.

For example, "pending" could represent someone who never actually finished a subscription, right? or even "trash" or "deleted".

btw, in some cases "on-hold" could be the status of a subscription that was never active, it just had a failed payment and never completed...

wdyt?

**Comment by leogermani** [comment]
> Also, do you think it's worth adding a unit test where we mock some users who own various types of products (and some who own none), to test the logic? Or is that overkill?

It's not overkill, I did some drafts but it takes a lot of mocking... and also it would basically require that we write tests for methods such as `get_active_subscriptions_for_user`, `get_most_recent_cancelled_or_expired_subscription` and `get_one_time_donation_order_for_user`... and all of them rely heavily on woocommerce plugins. I just thought it was too much effort for this one PR that needed to be shipped - and since the methods are small and simple enough, I didn't feel it was critical

File: tests/unit-tests/reader-activation-sync.php (hunk 2 above)
Comment (dkoo) [reply tests/unit-tests/reader-activation-sync.php]: I see, up to you! We can go back and add it in a separate PR too if we decide we want it

File: includes/reader-activation/sync/class-woocommerce.php (hunk 1 above)
Comment (dkoo) [reply includes/reader-activation/sync/class-woocommerce.php]: These are good points! In that case maybe it'd be a good idea to create a constant for `FORMER_SUBSCRIPTION_STATUSES` or something similar so that we can more easily explain why we're looking at these statuses in particular?

**Review by dkoo — APPROVED** [review]
I think it's okay to proceed without the extra tests—if we decide we do want them, that can be another PR in the future so they don't block this release.

File: includes/reader-activation/sync/class-woocommerce.php (hunk 1 above)
Comment (leogermani) [reply includes/reader-activation/sync/class-woocommerce.php]: https://github.com/Automattic/newspack-plugin/pull/3416/commits/e0b2778acb012af23faadbe67b924766e3928429

File: tests/unit-tests/reader-activation-sync.php (hunk 2 above)
Comment (leogermani) [reply tests/unit-tests/reader-activation-sync.php]: https://github.com/Automattic/newspack-plugin/pull/3416/commits/2f1a30551521f28ac7376826917fffc4c46375e4

**Review by dkoo — APPROVED** [review]
